File: problems/17_math_and_geometry/2013_detect_squares/solution.swift
// 2013. Detect Squares
import Foundation
class DetectSquares {
    var cnt: [Int: Int] = [:]
    var points: [(Int, Int)] = []
    func key(_ x: Int, _ y: Int) -> Int { x * 2001 + y }
    func add(_ point: [Int]) {
        let k = key(point[0], point[1])
        cnt[k, default: 0] += 1
        points.append((point[0], point[1]))
    }
    func count(_ point: [Int]) -> Int {
        let px = point[0], py = point[1]
        var total = 0
        for (x, y) in points {
            if abs(x - px) != abs(y - py) || x == px || y == py { continue }
            let a = cnt[key(x, py)] ?? 0
            let b = cnt[key(px, y)] ?? 0
            total += a * b
        }
        return total
    }
}
let ds = DetectSquares()
ds.add([3, 10]); ds.add([11, 2]); ds.add([3, 2])
if ds.count([11, 10]) != 1 { fatalError("a") }
if ds.count([14, 8]) != 0 { fatalError("b") }
ds.add([11, 2])
if ds.count([11, 10]) != 2 { fatalError("c") }
print("All tests passed!")

File: problems/17_math_and_geometry/2013_detect_squares/solution.py
"""
Problem: 2013. Detect Squares
"""
from collections import defaultdict
from typing import List


class DetectSquares:
    def __init__(self):
        self.cnt = defaultdict(int)
        self.points = []

    def add(self, point: List[int]) -> None:
        self.cnt[(point[0], point[1])] += 1
        self.points.append(point)

    def count(self, point: List[int]) -> int:
        px, py = point
        total = 0
        for x, y in self.points:
            if abs(x - px) != abs(y - py) or x == px or y == py:
                continue
            total += self.cnt[(x, py)] * self.cnt[(px, y)]
        return total


if __name__ == "__main__":
    ds = DetectSquares()
    ds.add([3, 10]); ds.add([11, 2]); ds.add([3, 2])
    assert ds.count([11, 10]) == 1
    assert ds.count([14, 8]) == 0
    ds.add([11, 2])
    assert ds.count([11, 10]) == 2
    print("All tests passed!")
